### PR TITLE
feat:Poll Images from ECR Container Repository Plugin

### LINKED ISSUE #3978: Feature: Polling Images from ECR Plugin
### 🔖 Feature description

User should be able to poll images from container registry and make it available in devtron system for deployments for ECR

### 🎤 Pitch

It would be used to poll from container images from ECR repository.

### ✌️ Solution

Plugin would be introduced which can be used to poll images from ECR Container repository

### 🔄️ Alternative

_No response_

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/devtron-labs/devtron/blob/main/CODE_OF_CONDUCT.md)

AB#5813

### PR DESCRIPTION
**Pull request by Shivam-nagar23**
# Description

Fixes #3978 

## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have performed a self-review of my own code.
* [ ] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: api/restHandler/app/BuildPipelineRestHandler.go [hunk 1]
```diff
@@ -400,7 +400,7 @@ func (handler PipelineConfigRestHandlerImpl) PatchCiPipelines(w http.ResponseWri
 		ciConfigRequest.DockerRegistry = emptyDockerRegistry
 		ciConfigRequest.AppId = patchRequest.AppId
 		ciConfigRequest.CiBuildConfig = &bean1.CiBuildConfigBean{}
-		ciConfigRequest.CiBuildConfig.CiBuildType = "skip-build"
+		ciConfigRequest.CiBuildConfig.CiBuildType = bean1.SKIP_BUILD_BUILD_TYPE
```
Comment (vikramdevtron) [inline api/restHandler/app/BuildPipelineRestHandler.go:403]: bean1.SKIP_BUILD_BUILD_TYPE >> bean1.SKIP_BUILD_TYPE

File: client/cron/CiTriggerCron.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	if err != nil {
+		logger.Errorw("error while configure cron job for ci workflow status update", "err", err)
+		return impl
+	}
+	return impl
+}
+
+type CiTriggerCronConfig struct {
+	SourceControllerCronTime int    `env:"CI_TRIGGER_CRON_TIME" envDefault:"2"`
+	PluginName               string `env:"PLUGIN_IDS"  envDefault:"Pull images from container repository"`
```
Comment (vikramdevtron) [inline client/cron/CiTriggerCron.go:55]: PluginName or PluginIds

File: util/helper.go [hunk 3]
```diff
@@ -323,3 +329,10 @@ func MatchRegexExpression(exp string, text string) (bool, error) {
 	matched := rExp.Match([]byte(text))
 	return matched, nil
 }
+
+func GetLatestImageAccToImagePushedAt(imageDetails []types.ImageDetail) types.ImageDetail {
+	sort.Slice(imageDetails, func(i, j int) bool {
+		return imageDetails[i].ImagePushedAt.After(*imageDetails[j].ImagePushedAt)
+	})
+	return imageDetails[0]
```
Comment (vikramdevtron) [inline util/helper.go:337]: what if imageDetails nil or empty

File: util/helper.go [hunk 4]
```diff
@@ -103,6 +105,10 @@ func ExtractChartVersion(chartVersion string) (int, int, error) {
 	return chartMajorVersion, chartMinorVersion, nil
 }
 
+func ExtractEcrImage(registryId, region, repoName, tag string) string {
+	return fmt.Sprintf("%s.dkr.ecr.%s.amazonaws.com/%s:%s", registryId, region, repoName, tag)
```
Comment (vikramdevtron) [inline util/helper.go:109]: put this string somewhere in constants

File: pkg/pipeline/CiCdPipelineOrchestrator.go [hunk 5]
```diff
@@ -803,7 +806,9 @@ func (impl CiCdPipelineOrchestratorImpl) CreateCiConf(createRequest *bean.CiConf
 
 		} else {
 			//save pipeline in db end
-			err = impl.AddPipelineMaterialInGitSensor(pipelineMaterials)
+			if len(pipelineMaterials) != 0 {
```
Comment (vikramdevtron) [inline pkg/pipeline/CiCdPipelineOrchestrator.go:809]: put more detailed comments

File: internal/sql/repository/CiArtifactRepository.go [hunk 6]
```diff
@@ -236,6 +239,36 @@ func (impl CiArtifactRepositoryImpl) GetArtifactsByCDPipeline(cdPipelineId, limi
...
+		"where cws.pipeline_id IN (" + helper.GetCommaSepratedString(ciPipelineIds) + "); "
+
+	_, err := impl.dbConnection.Query(&artifacts, query)
+	if err != nil {
+		return nil, err
+	}
+	return artifacts, nil
+}
+
+func (impl CiArtifactRepositoryImpl) GetLatestArtifactTimeByCiPipelineId(ciPipelineId int) (*CiArtifact, error) {
```
Comment (vikramdevtron) [inline internal/sql/repository/CiArtifactRepository.go:257]: put comments

File: client/cron/CiTriggerCron.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+	if err != nil || len(plugin) == 0 {
+		return
+	}
+
+	ciPipelineIds, err := impl.pipelineStageRepository.GetAllCiPipelineIdsByPluginIdAndStageType(plugin[0].Id, string(repository.PIPELINE_STAGE_TYPE_PRE_CI))
+	if err != nil {
+		return
+	}
+	//artifacts, err := impl.ciArtifactRepository.GetLatestArtifactTimeByCiPipelineIds(ciPipelineIds)
```
Comment (vikramdevtron) [inline client/cron/CiTriggerCron.go:81]: remove unwanted comments

File: util/helper.go (hunk 3 above)
Comment (Shivam-nagar23) [reply util/helper.go:337]: Have Already handled it where it is  used.

**Review by vikramdevtron — APPROVED** [review]

**Pull request merged**